### DISCUSSION TITLE
Does phalcon support attaching events to specific model before* or after* ?

### DISCUSSION THREAD

**Discussion by hostaway-greg** [phalcon/cphalcon#16541] [Q&A]
Hi,

I would like to attach external listener to specific model beforeSave and beforeUpdate hooks. Is it possible? I have tried putting FQN name into attach with beforeSave etc. but without any luck. I would like to avoid attaching it to all models - I have hundreds of them. If this is missing then it is a big omission IMHO. 

Thanks and have a nice day.

**Comment by niden** [comment]
You should be able to do that by using the methods in the model itself. Alternatively you can attach a listener with the before/after logic globally and have a simple `if` statement to filter for the model(s) you need

**Reply by hostaway-greg** [reply]
I would like to use external listener, some of my models handle different logic (like sending notifications, triggering some processing etc.) and having separate listeners is cleaner. Single responsibility principle. 
If I use generic db:beforeSave and start rolling events on more and more models, then that if check will be filtering hundreds of calls, and it will waste a lot of resources... best would be to be able to attach it just to one model... I can't believe it is not supported...

**Reply by s-ohnishi** [reply]
The Model has the following events, and if a method with the same name exists, it will be called at each timing.
Since it is already available, it may not be possible to add a new one.
- afterFetch
- prepareSave
- beforeValidation
- beforeValidationOnCreate
- validation
- afterValidationOnCreate
- afterValidation
- beforeSave
- beforeCreate
- afterCreate
- afterSave

So, I think the problem can be solved by providing a method only for the model you want to set.

**Comment by hostaway-greg** [comment]
> The Model has the following events, and if a method with the same name exists, it will be called at each timing. Since it is already available, it may not be possible to add a new one.
> So, I think the problem can be solved by providing a method only for the model you want to set.

I know about those methods. But as I mentioned before: in some of the bigger models I see that afterSave is getting bigger and bigger and there is a lot of different logic there. It would be much cleaner If I could attach multiple external listeners ONLY to specific model afterSave method. Not to all of them.

**Reply by s-ohnishi** [reply]
> ONLY to specific model afterSave method

I think it is possible to provide an afterSave method only for a specific model.

```PHP
class Foos extends \Phalcon\Mvc\Model
{
    public function afterSave()
    {
        // Required processing
    }
    ...
}
```

> Not to all of them.

Of course, this should only apply to models that have this method added.

**Reply by s-ohnishi** [reply]
If you have multiple models that have afterSave methods with the same content, I think they should just extend the class that defines the afterSave method.
However, I don't think they all process exactly the same thing.

**Comment by hostaway-greg** [comment]
OK, so in summary: there is no way to attach external listeners to a specific model hook. For example, there is no way to use `->attach("FQNModelName:beforeSave")` and have all logic external listener, it has to be inside FQNModel::beforeSave method. 
Thank you.

**Reply by s-ohnishi** [reply]
that's right.
Please try it.
